### ISSUE TITLE
how to split datasets ?

### ISSUE THREAD

**Issue by tfzxyinhao** [lyst/lightfm#179]
I know how to use lightfm, but I don't know how to split datasets when using it in first real case
is there have a documents show that?

**Comment by maciejkula** [comment]
You can have a look here for example: http://blog.ethanrosenthal.com/2016/10/19/implicit-mf-part-1/

**Comment by dofine** [comment]
Maybe not worth opening a new issue so I'll post here.  I read the examples in the stackexchange docs:

>The training and test set are divided chronologically:
>the test set contains the 10% of interactions that happened after the 90% in the training set.

And I've also read the code. Does this mean that all user-question interactions are sorted by time, and the last 10% of them are taken as test set? The train and test interaction matrix are of the same shape. But what confuses me is that, say a user `i` interacts with two questions `m` in the train set and `n` in the test set. We would have `train[i][m]=1` `train[i][n]=0` and `test[i][m]=0` `test[i][n]=1`. Do these two equations conflict with each other? How can a model, fitted with training set in whihch [i][n] = 0,   produce right recommendations in the test set?

I've read other tutorials such as [Implementing your own recommender systems in Python](https://cambridgespark.com/content/tutorials/implementing-your-own-recommender-systems-in-Python/index.html) and couldn't find an explanation. A newbie question. Hope you wouldn't mind. :)

**Comment by datascienceit** [comment]
+1

**Comment by maciejkula** [comment]
I've added splitting utilities in the master branch: https://github.com/lyst/lightfm/blob/master/lightfm/cross_validation.py#L20

@dofine sorry for the late response.
1. This is simply the nature of implicit feedback recommender systems. The model are trained and tuned to recommend those items with which a user does not interact in the training set but does interact with in the test set (likely, the recommendation score is lower than known (that is, present in training) items, but should still be high).
2. You could do this. Ultimately, you want your train/test split to reflect the conditions your production system is going to face.
3.  If features change you will probably want to have multiple rows per user (representing views over that user over time).

**Comment by AbderrahimAI** [comment]
For the second point, I think that the train_test_split function provided by lightFM doesn't split according to user but try to hide item from user in the training set. 
So how can I split my set so that a user won't appear in the train set at all?
